src/toast/ops/sim_gaindrifts.py

```diff
@@ -0,0 +1,322 @@
+# Copyright (c) 2015-2020 by the parties listed in the AUTHORS file.
+# All rights reserved.  Use of this source code is governed by
+# a BSD-style license that can be found in the LICENSE file.
+
+import traitlets
+
+import numpy as np
+from astropy import units as u
+from ..timing import function_timer
+
+from ..traits import trait_docs, Int, Float, Unicode, Bool, Quantity, Callable
+
+
+from .operator import Operator
+
+from ..utils import Environment, Logger
+from .. import rng
+from .sim_tod_noise import sim_noise_timestream
+
+
+@trait_docs
+class GainDrifter(Operator):
+    """Operator which injects gain drifts to the signal.
+
+    The drift can be injected into 3 different ways:
+    - `linear_drift`: inject a linear drift   with a random slope  for each detector
+    - `slow_drift`: inject a drift signal with a `1/f` PSD, simulated up to
+    the  frequencies<`cutoff_freq`, in case `cutoff_freq< (1/t_obs)`, `cutoff_freq=1/t_obs`.
+    - `thermal_drift`: inject a drift encoding frequencies up to the sampling rate, to simulate
+    the thermal fluctuations in the focalplane.
+    Both `slow_drift` and `thermal_drift` modes encode the possibility to inject a common mode drifts
+    to all the detectors belonging to a group of detectors identified the string `focalplane_group` ( can
+    be any string set by the user used to identify the groups in the detector table).
+    The amount of common mode contribution is set by setting detector_mismatch to a value `<1`, (with
+    0 being the case with only injecting common mode signal).
+
+    """
+
+    API = Int(0, help="Internal interface version for this operator")
+
+    det_data = Unicode(
+        "signal", help="Observation detdata key to inject the gain drift"
+    )
+
+    include_common_mode = Bool(
+        False, help="If True, inject a common drift to all the local detector group "
+    )
+
+    fknee_drift = Quantity(
+        20.0 * u.mHz,
+        help="fknee of the drift signal",
+    )
+    cutoff_freq = Quantity(
+        0.2 * u.mHz,
+        help="cutoff  frequency to simulate a slow  drift (assumed < sampling rate)",
+    )
+    sigma_drift = Float(
+        1e-3,
+        help="dimensionless amplitude  of the drift signal, (for `thermal_drift` corresponds to the thermal fluctuation level in K units)",
+    )
+    alpha_drift = Float(
+        1.0,
+        help="spectral index  of the drift signal spectrum",
+    )
+
+    detector_mismatch = Float(
+        1.0,
+        help="mismatch between detectors for `thermal_drift` and `slow_drift` ranging from 0 to 1. Default value implies no common mode injected",
+    )
+    thermal_fluctuation_amplitude = Quantity(
+        1 * u.K,
+        help="Amplitude of thermal fluctuation for `thermal_drift` in  Kelvin units ",
+    )
+    focalplane_Tbath = Quantity(
+        100 * u.mK,
+        help="temperature of the focalplane for `thermal_drift` ",
+    )
+    responsivity_function = Callable(
```

Comment:
This is a cool trait type that we have not used before.  It will be useful in many operators as a mechanism for user-provided callbacks!